### PR TITLE
fix(vim): allow space key in Vim insert mode inside editor (#88)

### LINKED ISSUE #88: Vim Space leader swallows the space key while typing
## Problem

When Vim mode is on, `globalKeys.ts` treats Space as a leader key.

`shouldHandleEvent` returns `true` inside CodeMirror, then Space calls `preventDefault()`.

That is fine in Normal mode. In Insert mode, the user is trying to type a space. The leader handler still eats it, so words stick together (`helloworld`) and a 1-second leader window starts.

## Suggested fix

- Only start the Space leader when the editor is in Normal / Visual mode, not Insert.
- Or ignore Space when the target is `.cm-content` and Vim is in insert.
- Add a tiny test or a manual checklist item: “Vim insert mode can type spaces”.

## Where to look

- `src/keybindings/globalKeys.ts`
- `src/editor/vimExtension.ts`
- `src/App.tsx` (where vim mode enables global keys)

## How to test

1. Turn on Vim mode.
2. Press `i`, type `hello world`.
3. There must be a space. Pressing Space in Normal mode can still open the leader.

### PR DESCRIPTION
**Pull request by 0xIta3hi**
## Description
Fixes #88
Prevents `globalKeys.ts` from intercepting the `Space` key as a leader key while the user is typing in Vim **Insert mode** inside CodeMirror.
### Changes:
- Listens to `oo:vim-mode-change` events in `src/keybindings/globalKeys.ts` to maintain current Vim mode state.
- Bypasses `Space` leader interception when keypress occurs inside `.cm-editor` and Vim mode is in `insert`.
- Added unit tests in `tests/global-keys.test.ts` verifying `Space` key behavior in Insert vs Normal mode.
## Verification
- [x] `npx vitest run tests/global-keys.test.ts` passed 5/5 tests.
- [x] `npm run lint` passed with 0 TypeScript compilation errors.

### CODE REVIEW

**Comment by 0xIta3hi** [comment]
@varshithm7x , Can you also look at this PR? if not just assign it to me back and i'll take care from there.

**Comment by rishi-jat** [comment]
will review this one @0xIta3hi

File: src/keybindings/globalKeys.ts [hunk 1]
```diff
@@ -83,6 +99,12 @@ function onGlobalKeydown(event: KeyboardEvent): void {
     !event.ctrlKey &&
     !event.metaKey
   ) {
+    const target = event.target as Element | null;
+    const withinCodeMirror = !!(target as HTMLElement | null)?.closest(".cm-editor");
+    if (withinCodeMirror && currentVimMode === "insert") {
```
Comment (rishi-jat) [inline src/keybindings/globalKeys.ts]: **[bug]** `currentVimMode` is a single module global updated only from `oo:vim-mode-change`. That event is not a reliable live signal of the focused editor. `toggleVimMode(view, true)` attaches the CM listener in a `queueMicrotask`, but `@replit/codemirror-vim` calls `enterVimMode()` synchronously in the ViewPlugin constructor and immediately signals `{mode: "normal"}`. That initial signal is always missed. Combined with the rest of the existing wiring, the cache is then wrong in two common cases: (1) User is in insert, switches notes — `Editor` destroys and recreates the view on `activePath` (`Editor.tsx` effect cleanup + remount) and the new editor starts in normal, but no `oo:vim-mode-change` is emitted, so `currentVimMode` stays `"insert"` and Space no longer starts the leader. (2) Toggling Vim off runs `setGlobalKeybindingsEnabled(false)` (which resets the cache to `"normal"`) and then `toggleVimMode(..., false)`, which dispatches `{mode: "INSERT"}`. `handleVimModeChange` still applies that event while keybindings are disabled, so re-enabling Vim leaves `currentVimMode === "insert"` until the user next changes modes. Split panes share the same global, so the last editor that changed mode also wins.

**Suggestion:** Do not trust a process-wide cache for this decision. At Space-handling time, resolve the focused CM view (e.g. `EditorView.findFromDOM(target)` + `getCM(view)?.state.vim`) and treat `insertMode` / replace as insert. If you keep the event cache, at least: ignore `oo:vim-mode-change` while `!enabled`; reset to `"normal"` on enable, not only on disable; and after `attachVimModeListener` hooks up, immediately publish the current CM mode so remounts and the missed `enterVimMode` signal cannot leave a stale `"insert"`.

Comment (rishi-jat) [inline src/keybindings/globalKeys.ts]: **[suggestion]** The bypass is only `currentVimMode === "insert"`. Issue #88 asked to start the Space leader in Normal/Visual, not every non-insert mode. The `:` command dialog is an `<input>` inside `.cm-editor` (`createVimPanel` / `openDialog`). `shouldHandleEvent` returns true for anything under `.cm-editor` before it ever checks `INPUT`, and `vimExtension.ts` reports that dialog as `"command"`. Space in `:find some file` is still `preventDefault()`’d. This was already broken before the PR; the new predicate is the place to close it.

**Suggestion:** Skip the Space leader unless the live mode is normal or visual (including `visual line` / `visual block`). Alternatively, stop treating vim dialog inputs as CodeMirror content in `shouldHandleEvent` (`INPUT`/`TEXTAREA` inside `.cm-editor` should not return true).

File: src/keybindings/globalKeys.ts [hunk 2]
```diff
@@ -83,6 +99,12 @@ function onGlobalKeydown(event: KeyboardEvent): void {
     !event.ctrlKey &&
     !event.metaKey
   ) {
+    const target = event.target as Element | null;
+    const withinCodeMirror = !!(target as HTMLElement | null)?.closest(".cm-editor");
+    if (withinCodeMirror && currentVimMode === "insert") {
+      return;
```
Comment (rishi-jat) [inline src/keybindings/globalKeys.ts]: **[suggestion]** The insert-mode early return leaves `leaderPending` and `leaderTimer` untouched. A Space that started the leader in normal/chrome, followed by a Space typed in insert within the 1s window, keeps the leader armed. The next character that is a leader key (`f`, `s`, `/`, …) is then stolen as a command instead of inserted. Uncommon, but this is a new hole: the handler used to always either consume Space as leader or fall through to the `leaderPending` follow-up logic.

**Suggestion:** Call `clearLeaderState()` before returning, or do not start/keep a leader sequence when the current key would be an insert-mode Space.

File: tests/global-keys.test.ts [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
   Object.defineProperty(event, "target", { value: target });
   return event;
 }
 
 describe("global key handling", () => {
+  beforeEach(() => {
+    initGlobalKeybindings();
+    setGlobalKeybindingsEnabled(true);
+    setVimMode("normal");
+  });
```
Comment (rishi-jat) [inline tests/global-keys.test.ts]: **[suggestion]** `beforeEach` adds the real `keydown` listener and can leave `leaderPending` / a 1s timer from the normal-mode Space test. Nothing resets leader state, and nothing dispatches `oo:vim-mode-change` — tests poke `setVimMode()` directly, so the event path Copilot flagged (and the remount/disable stale-cache bug) cannot fail CI. Current cases happen not to send a follow-up leader key, so they are not flaky today.

**Suggestion:** Reset with `setGlobalKeybindingsEnabled(false)` (or an exported `clearLeaderState`) at the start of each test, and add cases that set mode via `window.dispatchEvent(new CustomEvent("oo:vim-mode-change", { detail: { mode } }))`. A case that goes insert → (simulate remount / disable+enable) → Space in `.cm-editor` would have caught Issue 1.

**Comment by 0xIta3hi** [comment]
@varshithm7x Can you review and merge it?

**Comment by varshithm7x** [comment]
The logic successfully prevents global hotkeys from intercepting the Space key in Vim Insert mode, and the tests look solid.
I noticed one edge case when using split panes. If Editor A is in Normal mode and Editor B is in Insert mode, clicking back and forth between the two editors doesn't update the mode indicator or the global keybindings. This is because no `oo:vim-mode-change` event is dispatched when switching focus.

To fix this, we should track the `lastMode` per editor and dispatch it on `focusin`. Could you update `attachVimModeListener` inside `src/editor/vimExtension.ts` to do this? or something similar?


```
function attachVimModeListener(view: EditorView): void {
  const cm = getCM(view);
  if (!cm) return;

  const key = cm as unknown as object;
  if (boundModeListeners.has(key)) {
    return;
  }

  let lastMode = "normal";

  cm.on("vim-mode-change", (payload: VimModePayload) => {
    lastMode = normalizeModeForEvent(payload);
    dispatchOOEvent("oo:vim-mode-change", {
      mode: lastMode,
    });
  });

  cm.on("dialog", () => {
    const hasDialog = !!(cm as unknown as { state?: { dialog?: unknown } }).state?.dialog;
    lastMode = hasDialog ? "command" : "normal";
    dispatchOOEvent("oo:vim-mode-change", {
      mode: lastMode,
    });
  });

  // Dispatch the last mode whenever this editor gains focus
  view.dom.addEventListener("focusin", () => {
    dispatchOOEvent("oo:vim-mode-change", {
      mode: lastMode,
    });
  });

  boundModeListeners.add(key);
}
```

**Comment by varshithm7x** [comment]
@0xIta3hi work on this please

**Comment by 0xIta3hi** [comment]
referenced it in #118  @varshithm7x

**Comment by varshithm7x** [comment]
> referenced it in #118 @varshithm7x

#118 is closed

**Comment by varshithm7x** [comment]
@0xIta3hi updates?
